### PR TITLE
238 giving a fraction of samples instead of a number of samples in the subsample class

### PR DESCRIPTION
**Pull request by BaptisteCalot**
# Description

Fix addressing issue 238. We introduce the ability to create a training set using the split method of the subsample class with a fraction between 0 and 1 representing the proportion of the training set. The option to specify an integer representing the number of elements in the training set is still retained

## Type of change

- Bug fix 

In the case where the attribute self.n_samples is a float, the feature n_samples used in the split method of Subsample class becomes self.n_samples * X.shape[0], taking its floor integer part

# How Has This Been Tested?

- Test test_split_SubSample_n_samples

We check that the training and test sets build using the split method are as expected with the given seed. We use two instances of Subsample, one with an integer n_samples and the other with an n_samples less than 1

# Checklist

- [x] I have read the [contributing guidelines](https://github.com/simai-ml/MAPIE/blob/master/CONTRIBUTING.rst)
- [ ] I have updated the [HISTORY.rst](https://github.com/simai-ml/MAPIE/blob/master/HISTORY.rst) and [AUTHORS.rst](https://github.com/simai-ml/MAPIE/blob/master/AUTHORS.rst) files
- [x] Linting passes successfully : `make lint`
- [x] Typing passes successfully : `make type-check`
- [x] Unit tests pass successfully : `make tests`
- [x] Coverage is 100% : `make coverage`
- [x] Documentation builds successfully : `make doc`

### CODE REVIEW

File: HISTORY.rst [hunk 1]
```diff
@@ -5,6 +5,8 @@ History
 0.8.5 (2024-**-**)
```
Comment (LacombeLouis) [inline HISTORY.rst:5]: Note that this should be for 0.8.6.

File: mapie/subsample.py [hunk 2]
```diff
@@ -74,9 +75,12 @@ def split(
             The testing set indices for that split.
         """
         indices = np.arange(_num_samples(X))
-        n_samples = (
-            self.n_samples if self.n_samples is not None else len(indices)
-        )
+        if self.n_samples is None:
+            n_samples = len(indices)
+        elif isinstance(self.n_samples, float) and 0 < self.n_samples < 1:
```
Comment (LacombeLouis) [inline mapie/subsample.py:80]: I think this should raise an error no? Say it's 1.2?

**Review by LacombeLouis — CHANGES_REQUESTED** [review]
Thank you. I think we have some issues we want to clear up! But otherwise looks like a good PR. Note to also give a response to the issue #231!

File: mapie/subsample.py [hunk 3]
```diff
@@ -22,9 +22,10 @@ class Subsample(BaseCrossValidator):
     ----------
     n_resamplings : int
         Number of resamplings. By default ``30``.
-    n_samples: int
+    n_samples: float
```
Comment (thibaultcordier) [inline mapie/subsample.py:25]:
```suggestion
    n_samples: Union[int, float]
```

File: mapie/subsample.py (hunk 2 above)
Comment (thibaultcordier) [reply mapie/subsample.py:80]: I agree with this comment. Create a `check_n_samples_ratio` function in utils.py and create unit tests associated.

File: mapie/tests/test_subsample.py [hunk 4]
```diff
@@ -32,6 +32,31 @@ def test_split_SubSample() -> None:
     np.testing.assert_equal(tests, tests_expected)
 
 
+def test_split_SubSample_n_samples() -> None:
```
Comment (thibaultcordier) [inline mapie/tests/test_subsample.py:35]:
```suggestion
@pytest.mark.parametrize("n_samples", [0.8, 6])
def test_split_SubSample_n_samples(n_samples: Union[int, float]) -> None:
```
Prefer to use this convention when testing several values of a function.

Comment (thibaultcordier) [reply mapie/tests/test_subsample.py:35]: However, I like your test but it has the wrong name. Suggest a test called `test_n_samples_equivalence' which tests that using `n_samples` and `n_ratio=n_samples/n_all` produces the same split.

**Review by thibaultcordier — CHANGES_REQUESTED** [review]
Thank you for your PR. I suggest you write separate unit tests for better code maintenance.

File: mapie/subsample.py [hunk 5]
```diff
@@ -74,9 +76,14 @@ def split(
...
         """
         indices = np.arange(_num_samples(X))
-        n_samples = (
-            self.n_samples if self.n_samples is not None else len(indices)
-        )
+        if self.n_samples is None:
+            n_samples = len(indices)
+        else:
+            n_samples = check_n_samples(X, self.n_samples)
+        # elif isinstance(self.n_samples, float) and 0 < self.n_samples < 1:
```
Comment (LacombeLouis) [inline mapie/subsample.py:83]: I don't think this is needed anymore?

File: mapie/subsample.py [hunk 6]
```diff
@@ -74,9 +76,14 @@ def split(
             The testing set indices for that split.
         """
         indices = np.arange(_num_samples(X))
-        n_samples = (
-            self.n_samples if self.n_samples is not None else len(indices)
-        )
+        if self.n_samples is None:
```
Comment (LacombeLouis) [inline mapie/subsample.py:79]: Why isn't this included in `check_n_samples`?

File: mapie/tests/test_utils.py [hunk 7]
```diff
@@ -508,3 +509,18 @@ def test_check_no_agg_cv_value_error(cv: Any) -> None:
         match=r"Allowed values must have the `get_n_splits` method"
     ):
         check_no_agg_cv(X_toy, cv, array)
+
+
+@pytest.mark.parametrize("X", [1, 2, 3, 4, 5, 6, 7, 8, 9, 10])
```
Comment (LacombeLouis) [inline mapie/tests/test_utils.py:514]: No need to test for so many values. Keep it simpler in this case. There is a choice to make between a bunch of tests and the time it takes to run them.

File: mapie/utils.py [hunk 8]
```diff
@@ -1332,3 +1332,39 @@ def check_arrays_length(*arrays: NDArray) -> None:
...
+        n_samples
+
+    Raises
+    ------
+    ValueError
+        If n_samples is not an int or a float between 0 and 1.
+    """
+    if isinstance(n_samples, float) and 0 < n_samples < 1:
+        n_samples = int(np.floor(n_samples * X.shape[0]))
+    elif isinstance(n_samples, float) and n_samples > 1:
```
Comment (LacombeLouis) [inline mapie/utils.py:1364]: What if `n_samples=-0.5`? Why are we only checking for `n_samples` greater than 1?

File: mapie/utils.py [hunk 9]
```diff
@@ -1332,3 +1332,39 @@ def check_arrays_length(*arrays: NDArray) -> None:
         raise ValueError(
                 "There are arrays with different length"
             )
+
+
+def check_n_samples(
```
Comment (LacombeLouis) [inline mapie/utils.py:1337]: What happens if `n_samples` is exactly `0` or `1`?

**Review by LacombeLouis — CHANGES_REQUESTED** [review]

**Review by LacombeLouis — CHANGES_REQUESTED** [review]
Hey @BaptisteCalot,
Thank you for this PR! I think there are still some adjustments to make! As mentioned in issue #231, I think it's important to have a default value.
Thank you!

File: mapie/tests/test_subsample.py [hunk 10]
```diff
@@ -32,6 +32,31 @@ def test_split_SubSample() -> None:
...
+    test1 = np.concatenate([x[1] for x in cv1.split(X)])
+    train2 = np.concatenate([x[0] for x in cv2.split(X)])
+    test2 = np.concatenate([x[1] for x in cv2.split(X)])
+    train1_expected = np.array([2, 8, 4, 9, 1, 6, 7,
+                                3, 3, 5, 1, 2, 9, 8, 0, 6])
+    test1_expected = np.array([0, 5, 4, 7])
+    train2_expected = np.array([2, 8, 4, 9, 1, 6, 3, 5, 1, 2, 9, 8])
+    test2_expected = np.array([0, 3, 5, 7, 0, 4, 6, 7])
+    expected_n_samples_cv1 = int(np.floor(0.8 * X.shape[0]))
+    assert len(train1) == 2 * expected_n_samples_cv1
```
[the remark this answers is not in the document]
Comment (BaptisteCalot) [reply mapie/tests/test_subsample.py]: I have taken it into account

File: mapie/subsample.py (hunk 3 above)
Comment (BaptisteCalot) [reply mapie/subsample.py:26]: It is now Optional[Union[int, float]]

File: mapie/tests/test_utils.py [hunk 11]
```diff
@@ -508,3 +510,19 @@ def test_check_no_agg_cv_value_error(cv: Any) -> None:
         match=r"Allowed values must have the `get_n_splits` method"
     ):
         check_no_agg_cv(X_toy, cv, array)
+
+
+@pytest.mark.parametrize("n_samples", [-5.5, -4, 0, 1.2])
```
[the remark this answers is not in the document]
Comment (BaptisteCalot) [reply mapie/tests/test_utils.py]: Thank you for your insightful comment Louis, I have taken it into account

File: mapie/utils.py [hunk 12]
```diff
@@ -1355,3 +1355,50 @@ def check_arrays_length(*arrays: NDArray) -> None:
...
+                "are float in the range (0.0, 1.0) or"
+                " int in the range [1, inf)"
+                )
+    elif isinstance(n_samples, int) and n_samples <= 0:
+        raise ValueError(
+                "Invalid n_samples. Allowed values "
+                "are float in the range (0.0, 1.0) or"
+                " int in the range [1, inf)"
+            )
+    return int(n_samples)
```
[the remark this answers is not in the document]
Comment (LacombeLouis) [reply mapie/utils.py:1410]: So, when `replace=False`, what's the error?

**Review by LacombeLouis — APPROVED** [review]
Hey @BaptisteCalot,
Perfect, thank you! Please quickly mention the error that will be obtained when `replace=False`. Otherwise, I think this can be validated today!

File: mapie/tests/test_utils.py [hunk 13]
```diff
@@ -508,3 +509,67 @@ def test_check_no_agg_cv_value_error(cv: Any) -> None:
...
+        match=re.escape(
+            r"Invalid n_samples. Allowed values "
+            r"are float in the range (0.0, 1.0) or"
+            r" int in the range [1, inf)"
+        )
+    ):
+        check_n_samples(X=X, n_samples=n_samples, indices=indices)
+
+
+@pytest.mark.parametrize("n_samples", [1.2, 2.5, 3.4])
```
Comment (LacombeLouis) [inline mapie/tests/test_utils.py:562]: You can put this test with the `test_invalid_n_samples_float_negative`. Good to separate, but this is a bit too much 😅. (So simply get rid of this test and add [-5.5, -4.3, -0.2, 1.2, 3.64]

File: mapie/utils.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (LacombeLouis) [reply mapie/utils.py:1410]: So, when `replace=False`, what's the error?

File: mapie/tests/test_utils.py (hunk 13 above)
Comment (BaptisteCalot) [reply mapie/tests/test_utils.py:562]: Okay, I have just taken it into account

File: mapie/utils.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (BaptisteCalot) [reply mapie/utils.py:1410]: When replace = False, an error is raised by the resample function from scikit-learn that we use in the split method of Subsample : [link](https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/utils/_indexing.py#L515)

File: mapie/utils.py [hunk 14]
```diff
@@ -1355,3 +1355,56 @@ def check_arrays_length(*arrays: NDArray) -> None:
...
+                "Invalid n_samples. Allowed values "
+                "are float in the range (0.0, 1.0) or"
+                " int in the range [1, inf)"
+                )
+    elif isinstance(n_samples, int) and n_samples <= 0:
+        raise ValueError(
+                "Invalid n_samples. Allowed values "
+                "are float in the range (0.0, 1.0) or"
+                " int in the range [1, inf)"
+            )
```
Comment (thibaultcordier) [inline mapie/utils.py:1409]:
```suggestion
        raise ValueError(
             "Invalid n_samples. Allowed values "
             "are float in the range (0.0, 1.0) or"
             " int in the range [1, inf)"
             )
```

**Review by thibaultcordier — APPROVED** [review]
